### PR TITLE
Multithreaded implementation

### PR DESCRIPTION
**Pull request by netj**
Lays down the foundation for #3.  This version will be used for fixing crashes on latest Mac (10.11) caused by nonblocking I/O implementation (HazyResearch/deepdive#522).

@alldefector can you take a look at the code, perhaps just the [mkmimo_multithreaded.c](https://github.com/netj/mkmimo/pull/8/files#diff-a1e439a448f4775c4138223c606c8ce0) if the rest seems too noisy?

### CODE REVIEW

File: mkmimo_multithreaded.c [hunk 1]
```diff
@@ -1,15 +1,310 @@
...
+      int num_bytes_readable = buf->capacity - buf->size;
+      DEBUG("%s: can read %d bytes", input->name, num_bytes_readable);
+
+      int num_bytes_read = read(input->fd, buf->data + buf->begin + buf->size,
+                                num_bytes_readable);
+      DEBUG("%s: %d bytes read", input->name, num_bytes_read);
+
+      if (num_bytes_read < 0) {
+        // Close input upon errors
+        perrorf("read %s", input->name);
```
Comment (alldefector) [inline mkmimo_multithreaded.c]: Message is not clear that there is an IO error. Maybe also print out `num_bytes_read` if the code is meaningful.

**Comment by alldefector** [comment]
Looks great! How does its performance compare with nonblocking?

File: mkmimo_multithreaded.c (hunk 1 above)
Comment (netj) [inline mkmimo_multithreaded.c]: `perrorf` is a macro that appends the actual error string that corresponds to the last error number, so I think it'll look clear.  Actually, I realize two issues:
1. I think it currently relies on errno global, so I probably need to make it reentrant/threadsafe.
2. Actually these errors are not escalating into a non-zero exit status, then will surely get silenced, which will be really awful.

Thanks for point me to this direction :)

**Comment by netj** [comment]
@alldefector Thanks for taking a look!  Haven't done any real measurement, but it easily passes throughput tests on Mac, which wasn't the case for nonblocking impl.  So I'll handle the error exit status and also measure things a bit.  I'm hopeful this will saturate system's data bandwidth, e.g., ~100GB/s for raiders as I measured before (#3).

Btw while coding this, I was running nonblocking impl. on an iMac with 10.10, keeping it busy at 300-400MB/s, but it never rebooted once.  Hilariously, my laptop rebooted while writing this comment as I was waiting for the first deepdive make test to finish on my fresh MacBook Pro.. and I'm now typing it again from scratch.  We need to finish this asap and fix deepdive.. :)

**Comment by netj** [comment]
`errno` turns out to be already thread-safe, and I made sure the code exits with non-zero status upon I/O errors but failed to create a precise test.

When I tried a very rough throughput test with something like below, the new multithreaded showed 2-5x improvement in throughput, each channel going 40MB/s (nonblocking) -> 100-250MB/s (multithreaded) for 30 I/O streams.  But there are lots of knobs to turn here, so not clear where these two shine.  @sashaperigo and I will start exploring this next quarter.  Anyway, the conclusion is the new implementation often delivers higher throughput, doesn't crash Mac. :)

``` bash
xMKMIMO_IMPL=nonblocking \
MULTIBUFFERING=2 \
PATH=$PATH:$PWD/test/util \
mkmimo_throughput timeout=1m \
num_producer_fast=50 \
num_producer_slow=10 \
num_consumer_fast=50 \
num_consumer_slow=10 \
producer_fast_throughput= \
producer_slow_throughput=2k \
consumer_fast_throughput= \
consumer_slow_throughput=1k \
#
```

One last thing that concerns me is a potential deadlock issue.  I observed tests hanging either at the beginning or at the end, and I think I'll have to do some stress tests to make sure it's completely reliable.  If you can, please take a look at some obvious parts that play with mutex and embarrass me. :)

**Comment by alldefector** [comment]
Seems good to go!

**Pull request merged**